tests/qos/buffer_helpers.py

```diff
@@ -41,3 +51,64 @@ def update_db_info(self):
         self.config_db = self.get_config_db()
         self.appl_db = self.get_appl_db()
         self.asic_db = self.get_asic_db()
+
+
+def get_ports_with_config_exceed_max_headroom(duthost):
+    config_db_info = DutDbInfo(duthost).get_config_db()
+    map_port_to_cable_len = config_db_info.get("CABLE_LENGTH|AZURE").get("value")
+    ports_with_config_exceed_max_headroom_ports = {}
+    speed_cable_len_exceed_max_headroom = {"speed": 400000, "cable_len": 200}
+
+    for port, cable_len in map_port_to_cable_len.items():
+        port_speed = config_db_info.get(f"PORT|{port}").get("value").get("speed")
+        if port_speed and cable_len:
+            if int(port_speed) >= speed_cable_len_exceed_max_headroom["speed"] and\
+                    int(cable_len.split("m")[0]) >= speed_cable_len_exceed_max_headroom["cable_len"]:
+                ports_with_config_exceed_max_headroom_ports.update({port: cable_len})
+    return ports_with_config_exceed_max_headroom_ports, map_port_to_cable_len
+
+
+def change_ports_cable_len(duthost, port_cable_info):
+    ports_cable_len_j2_file_name = "ports_cable_len.j2"
+    with open(os.path.join(PORT_CABLE_LEN_JSON_TEMPLATE_PATH, ports_cable_len_j2_file_name)) as template_file:
+        t = Template(template_file.read())
+
+    content = t.render(ports_cable_info=port_cable_info, ports_cable_info_len=len(port_cable_info))
+    logger.info(f"port cable len json content is {port_cable_info}")
+    ports_cable_len_config_json_file_name = "ports_cable_len_config.json"
+
+    cmd_gen_port_cable_len_config = f"cat << EOF >  {ports_cable_len_config_json_file_name} \n {content}"
+
+    duthost.shell(cmd_gen_port_cable_len_config)
+    duthost.shell("sudo config load {} -y".format(ports_cable_len_config_json_file_name))
+
+
+def gen_ports_cable_info(ports_with_config_exceed_max_headroom_ports, map_port_to_cable_len, updated_cable_len):
+    ports_cable_info = copy.deepcopy(map_port_to_cable_len)
+    if updated_cable_len:
+        for port, cable_len in ports_cable_info.items():
+            if port in ports_with_config_exceed_max_headroom_ports:
+                ports_cable_info[port] = updated_cable_len
+
+    return ports_cable_info
+
+
+@pytest.fixture(scope="function")
+def update_cable_len_for_all_ports(duthost):
+    if is_mellanox_device(duthost):
+        ports_cable_len_map_with_config_exceed_max_headroom, original_ports_cable_len_map = \
+            get_ports_with_config_exceed_max_headroom(duthost)
+
+        if ports_cable_len_map_with_config_exceed_max_headroom:
```

Comment:
Why only for 400G ports? How about other speed ports?

Reply:
Because we only encounter the issue on 400G. If we have other cases, we can update the dict of speed_cable_len_exceed_max_headroom  later